src/types/array.ts

```diff
@@ -1,16 +1,21 @@
 import { Runtype, Static, create } from '../runtype';
 import { ValidationError } from '../errors';
 
-interface Arr<E extends Runtype> extends Runtype<Static<E>[]> {
+type ArrayStaticType<E extends Runtype, RO extends boolean | undefined> = RO extends true ? ReadonlyArray<Static<E>> : Static<E>[]
+
+interface Arr<E extends Runtype, RO extends boolean | undefined> extends Runtype<ArrayStaticType<E, RO>> {
   tag: 'array';
   element: E;
+  isReadonly: RO;
+
+  asReadonly(): Arr<E, true>;
 }
 
 /**
  * Construct an array runtype from a runtype for its elements.
  */
-function Arr<E extends Runtype>(element: E): Arr<E> {
-  return create<Arr<E>>(
+function Arr<E extends Runtype, RO extends boolean | undefined = false>(element: E, isReadonly?: RO): Arr<E, RO> {
```

Comment:
Is the `| undefined` necessary?

Reply:
Actually now that I tested it, looks like it isn't required. I had the impression it would be necessary because of the isReadonly parameter is optional

Reply:
Just using classic function overloading seems to be the best solution actually